java-checks/src/main/java/org/sonar/java/checks/StringConcatToTextBlockCheck.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012-2021 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import org.sonar.check.Rule;
+import org.sonar.java.JavaVersionAwareVisitor;
+import org.sonar.java.model.ExpressionUtils;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.JavaVersion;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.Tree;
+
+@Rule(key = "S6126")
+public class StringConcatToTextBlockCheck extends IssuableSubscriptionVisitor implements JavaVersionAwareVisitor {
+
+  private static final String MESSAGE = "Replace this String concatenation with Text block.";
+  private static final Set<Tree> VISITED_NODES = new HashSet<>();
+
+  @Override
+  public boolean isCompatibleWithJavaVersion(JavaVersion version) {
+    return version.isJava15Compatible();
+  }
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return Collections.singletonList(Tree.Kind.PLUS);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    VISITED_NODES.add(tree);
+    if (VISITED_NODES.contains(tree.parent())) {
+      return;
+    }
```

Comment:
A nice trick to prevent reentrance.
I just wonder if we should fixe this case of issue duplication:
```
    System.out.println("<html>\n" + // Noncompliant, 1st issue
                "    <body>\n" +
                (
                "        <tag>\n" + // <== 2nd issue here
                "          xxx\n" +
                "        </tag>\n"
                ) +
                "    </body>\n" +
                "</html>");
```

Reply:
I think yes, we should report once here